### ISSUE TITLE
Exception when loading data into server

### ISSUE THREAD

**Issue by johnstairs** [microsoft/fhir-server#1353]
I was loading data into the FHIR server (16 concurrent requests) and got this exception:

```
System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.ContainsKey(TKey key)
   at Microsoft.Health.Fhir.Core.Features.Definition.SearchParameterDefinitionManager.UpdateSearchParameterHashMap(Dictionary`2 updatedSearchParamHashMap) in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Core\Features\Definition\SearchParameterDefinitionManager.cs:line 124
   at Microsoft.Health.Fhir.Core.Features.Definition.SearchParametersHashUpdatedHandler.Handle(SearchParametersHashUpdated notification, CancellationToken cancellationToken) in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Core\Features\Definition\SearchParametersHashUpdatedHandler.cs:line 26
   at MediatR.Internal.NotificationHandlerWrapperImpl`1.<>c__DisplayClass0_0.<Handle>b__1(INotification theNotification, CancellationToken theToken)
   at MediatR.Mediator.PublishCore(IEnumerable`1 allHandlers, INotification notification, CancellationToken cancellationToken)
   at Microsoft.Health.Fhir.Core.Features.Search.Registry.SearchParameterStatusManager.EnsureInitialized() in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Core\Features\Search\Registry\SearchParameterStatusManager.cs:line 145
   at Microsoft.Health.Api.Modules.InitializationModule.<>c__DisplayClass2_0.<<Configure>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Health.Fhir.Api.Features.Context.FhirRequestContextAfterAuthenticationMiddleware.Invoke(HttpContext context, IFhirRequestContextAccessor fhirRequestContextAccessor) in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Api\Features\Context\FhirRequestContextAfterAuthenticationMiddleware.cs:line 36
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.Health.Fhir.Api.Features.Context.FhirRequestContextBeforeAuthenticationMiddleware.Invoke(HttpContext context) in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Api\Features\Context\FhirRequestContextBeforeAuthenticationMiddleware.cs:line 45
   at Microsoft.Health.Fhir.Api.Features.ApiNotifications.ApiNotificationMiddleware.PublishNotificationAsync(HttpContext context, RequestDelegate next) in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Api\Features\ApiNotifications\ApiNotificationMiddleware.cs:line 63
   at Microsoft.Health.Fhir.Api.Features.ApiNotifications.ApiNotificationMiddleware.PublishNotificationAsync(HttpContext context, RequestDelegate next) in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Api\Features\ApiNotifications\ApiNotificationMiddleware.cs:line 90
   at Microsoft.Health.Fhir.Api.Features.ApiNotifications.ApiNotificationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) in D:\repos\fhir-server\src\Microsoft.Health.Fhir.Api\Features\ApiNotifications\ApiNotificationMiddleware.cs:line 52
```

**Comment by brendankowitz** [comment]
Is this on startup? The stack trace might suggest that `SearchParameterStatusManager.EnsureInitialized()` is being called multiple times

**Comment by johnstairs** [comment]
The exception was not on startup. I had been loading data for some minutes.

**Comment by namadabu** [comment]
There is a fix out for handling concurrent calls to `SearchParameterDefinitionManager.UpdateSearchParameterHashMap` in the above PR. Theoretically this situation should arise only if `EnsureInitialized` is called twice on the `SearchParameterStatusManager` (which is not expected). I tried to repro this issue locally in order to understand the root cause. I ingested data at the rate of ~16 concurrent requests (using both CosmosDb and SQL as backing data stores) multiple times but did not run into this exception. I also checked our test and release environment logs and haven't seen this exception.
